### DISCUSSION TITLE
Coalesce Optional Query Param to Nil

### DISCUSSION THREAD

**Discussion by heyltsjay** [pointfreeco/swift-url-routing#18] [Ideas]
In the following example I would expect some way to coalesce `Field("password")` to nil without `Optionally`, but it seems like there is no way to distinguish `nil` as being the default value itself, rather than the absence of one.  I was surprised the compiler was happy with the pattern but that the tests were failing to parse `app://signIn?email=foo@pointfree.co`

```
public extension AppRoute {
    
    enum Auth: Equatable {
        case signIn(email: String, password: String?)
    }

    static let authRouter = OneOf {
        Route(.case(Auth.signIn)) {
            Path { "signIn" }
            Query {
                Field("email")
                Optionally {
                    Field("password")
                }
            }
        }
    }

}
```

**Comment by stephencelis** [comment]
@heyltsjay I think this is working as intended. All parser optionality is handled explicitly via `Optionally`, and the `Field("name")` initializer is a shorthand for `Field("name", .string)`.

It might be possible that another overload could be introduced, but we're not sure result builder type inference and/or type ambiguity would interfere. For example, if it weren't for your `Route` mapping the `String?` to an enum's associated value, what kind of parser would this be in isolation?

```swift
let p = Field("email") // Parser<Substring, ???>
```

Some magic use of `@_disfavoredOverload` could work, and if you explore it by defining an overload of `Field`, please do let us know what you find, and we'd definitely consider a PR if such an overload doesn't cause other problems.

**Comment by stephencelis** [comment]
I'm also going to move this to a discussion since it's more about discussing a potential feature than a bug in the library.

**Comment by heyltsjay** [comment]
This all sounds fair enough, albeit breaking my mental model somewhat of how other optional query params are treated.

**Reply by heyltsjay** [reply]
The nuts and bolts of keeping the compiler happy is probably too academic for me, but I am a little surprised that it is happy to match against the optional whereas I don't think there is an outcome that would write `nil` without throwing an error.
